gateway/app/drivers/quality_on_demand/nef/nef_interface.py

```diff
@@ -0,0 +1,403 @@
+        self, body: UserPlaneNotificationData, id: str
+    ) -> None:
+        sub_info = await self.get_qod_information_by_id(id)
+
+        if sub_info is None:
+            raise ResourceNotFound()
+
+        status = self._get_qod_status(body)
+        if status is not None:
+            sub_info.qosStatus = status
+
+        if sub_info.qosStatus == Status.UNAVAILABLE:
+            sub_info.statusInfo = StatusInfo.NETWORK_TERMINATED
+
+        sink = sub_info.sink
+        if sink is None:
+            return
+
+        cloud_event = CloudEvent(
+            id=str(uuid.uuid4()),
+            source=self.source,
+            type=NotificationEventType.org_camaraproject_quality_on_demand_v1_qos_status_changed,
+            specversion=Specversion.field_1_0,
+            datacontenttype=Datacontenttype.application_json,
+            data=Data(
+                sessionId=sub_info.sessionId,
+                qosStatus=EventQosStatus.AVAILABLE
+                if sub_info.qosStatus == Status.AVAILABLE
+                else EventQosStatus.UNAVAILABLE,
+                statusInfo=sub_info.statusInfo,
+            ),
+            time=datetime.datetime.now(),
+        )
+
+        await self.httpx_client_callback.post(
+            str(sink),
+            content=cloud_event.model_dump_json(exclude_unset=True, by_alias=True),
+        )
+
+        await self._save_subscription_info(id, sub_info)
+
+    def _get_qod_status(self, nef_response: UserPlaneNotificationData) -> Status | None:
+        status = None
+        for report in nef_response.eventReports:
+            if report.event == UserPlaneEvent.QOS_GUARANTEED:
+                status = Status.AVAILABLE
+            elif report.event == UserPlaneEvent.FAILED_RESOURCES_ALLOCATION:
+                status = Status.UNAVAILABLE
+            elif report.event == UserPlaneEvent.QOS_NOT_GUARANTEED:
+                status = Status.REQUESTED
+            elif report.event == UserPlaneEvent.SUCCESSFUL_RESOURCES_ALLOCATION:
+                status = Status.AVAILABLE
+        return status
+
+    async def _save_subscription_info(self, id: str, data: SessionInfo) -> None:
+        key = f"{_prefix_info}:{id}"
+
+        await self.redis.set(
+            key, data.model_dump_json(exclude_unset=True, by_alias=True)
+        )
+
+    def _get_flow_info(self, req: BaseSessionInfo) -> List[FlowInfo]:
+        flow = []
+
+        device_ports = ""
+        if req.devicePorts:
+            device_ports = self._format_ports(req.devicePorts)
+
+        server_ports = ""
+        if req.applicationServerPorts:
+            server_ports = self._format_ports(req.applicationServerPorts)
+
+        dev_ips = []
+        app_ips = []
+
+        if req.device and req.device.ipv4Address is not None:
+            dev_ips.append(str(req.device.ipv4Address.publicAddress))
+
+        if req.applicationServer.ipv4Address is not None:
+            app_ips.append(req.applicationServer.ipv4Address)
```

Comment:
This need to be one if, you only add the address family if both the device and application server have it.